### PR TITLE
79 custom error pages

### PR DESCRIPTION
**Pull request by gwolf-011235**
### Summary
1. Support for custom error pages
2. Added sample custom error pages to static website
3. Fix: redirect from TargetResourceHandler is stored in Location Header
4. Small things

### 1. Support custom error pages
- ResponseBodyHandler checks if it finds a custom error page.
- Implemented logic to fetch custom one and use it instead of the default ones
- Added tests

### 2. Static website
- Static website has new folder error with two error pages:
  - 404 File not found
  - 403 Forbidden 

### 3. Fix
- In case of a directory which is accessed without '/' an automatic redirect is initiated
- Beforehand this was stored in targetResource, which caused problems: it got overwritten by ResponseBodyHandler
- Now is stored in Location header, which gets inserted. Since this header is only used in response, even an existing one can be safely overwritten

### 4. Small things
- Added configuration entries for custom error pages in the dummy configuration
- Added new status code 308 Permanent Redirect (not used right now, but delimits 3xx error codes)
- Moved the declaration for `getDefaultErrorPage` from utilities to `ResponseBodyHandler`. Also added "test" to increase code coverage -> management move!
- Added GTEST_SKIP() to failing ConfigFileParse test, so that unittest returns with no error

### CODE REVIEW

**Review by u413-284-si — CHANGES_REQUESTED** [review]
Looks good, like the custom pages! 

One minor comment, regarding passed parameters.

File: inc/ResponseBodyHandler.hpp [hunk 1]
```diff
@@ -14,13 +16,18 @@
...
-	explicit ResponseBodyHandler(HTTPRequest& request, std::string& responseBody, const FileSystemPolicy& fileSystemPolicy);
+	explicit ResponseBodyHandler(
+		Connection& connection, std::string& responseBody, const FileSystemPolicy& fileSystemPolicy);
 	void execute();
 
 private:
 	void handleErrorBody();
+	void setDefaultErrorPage();
 
+	Connection& m_connection;
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply inc/ResponseBodyHandler.hpp:27]: m_request is a refrence to connection.m_request.
It's there to shorten every `connection.m_request` to just `m_request`:

Eg:
```
[....]
if (m_request.httpStatus >= StatusMovedPermanently) {
		handleErrorBody();
	} else if (m_request.hasAutoindex) {
		AutoindexHandler autoindexHandler(m_fileSystemPolicy);
		m_responseBody = autoindexHandler.execute(m_request.targetResource);
		if (m_responseBody.empty()) {
			m_request.httpStatus = StatusInternalServerError;
			handleErrorBody();
			return;
		}
		m_request.httpStatus = StatusOK;
		m_request.targetResource += "autoindex.html";
[...]
```

vs:
```
[...]
if (connection.m_request.httpStatus >= StatusMovedPermanently) {
		handleErrorBody();
	} else if (connection.m_request.hasAutoindex) {
		AutoindexHandler autoindexHandler(m_fileSystemPolicy);
		m_responseBody = autoindexHandler.execute(connection.m_request.targetResource);
		if (m_responseBody.empty()) {
			connection.m_request.httpStatus = StatusInternalServerError;
			handleErrorBody();
			return;
		}
		connection.m_request.httpStatus = StatusOK;
		connection.m_request.targetResource += "autoindex.html";
[...]
```

Makes it more readable, don't you think?

File: src/ResponseBodyHandler.cpp [hunk 2]
```diff
@@ -3,12 +3,14 @@
...
+ * @param connection The Connection for which the response body is handled.
  * @param responseBody Saves the response body.
  * @param fileSystemPolicy File system policy. Can be mocked if needed.
  */
-ResponseBodyHandler::ResponseBodyHandler(HTTPRequest& request, std::string& responseBody, const FileSystemPolicy& fileSystemPolicy)
-	: m_request(request)
+ResponseBodyHandler::ResponseBodyHandler(
+	Connection& connection, std::string& responseBody, const FileSystemPolicy& fileSystemPolicy)
+	: m_connection(connection)
+	, m_request(connection.m_request)
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/ResponseBodyHandler.cpp:13]: See above

File: src/ResponseBodyHandler.cpp [hunk 3]
```diff
@@ -51,13 +51,182 @@ void ResponseBodyHandler::execute()
...
+		return;
+	}
+
+	LOG_DEBUG << "Custom error page: " << iter->second;
+
+	statusCode oldStatus = m_request.httpStatus;
+	m_request.httpStatus = StatusOK;
+	m_request.uri.path = iter->second;
+	TargetResourceHandler targetResourceHandler(m_fileSystemPolicy);
+	targetResourceHandler.execute(m_connection, m_request);
```
[the remark this answers is not in the document]
Comment (gwolf-011235) [reply src/ResponseBodyHandler.cpp:85]: This would mean a refactor of `TargetResourceHandler::execute()`.

This is done in #87

**Review by QCHR1581 — APPROVED** [review]
Cool custom pages! Appreciate the Rickroll :)

Nothing additional from my side.

File: inc/ResponseBodyHandler.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply inc/ResponseBodyHandler.hpp:27]: If you prefer, I am also fine with it. Not an important point

File: src/ResponseBodyHandler.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (u413-284-si) [reply src/ResponseBodyHandler.cpp:85]: Yep, saw that afterwards when reviewing #87

**Review by u413-284-si — APPROVED** [review]

**Pull request merged**